pandas/core/categorical.py

```diff
@@ -316,13 +317,14 @@ def __init__(self, values, categories=None, ordered=None, dtype=None,
             if not isinstance(values, np.ndarray):
                 values = _convert_to_list_like(values)
                 from pandas.core.series import _sanitize_array
-                # On list with NaNs, int values will be converted to float. Use
-                # "object" dtype to prevent this. In the end objects will be
-                # casted to int/... in the category assignment step.
-                if len(values) == 0 or isna(values).any():
+                # By convention, empty lists result in object dtype:
+                if len(values) == 0:
                     sanitize_dtype = 'object'
                 else:
                     sanitize_dtype = None
+                null_mask = isna(values)
```

Comment:
I think you can move the 

```
null_mask = isna(values)
```
to line 291 (where you set it to: ``null_mask`` = np.array(False)``

as we *always* need to check this (whether its an array or list-like anyhow)

Reply:
> as we always need to check this (whether its an array or list-like anyhow)

Well, no: conceptually, because if the user passes an ``array`` (or a pandas object) with missing values, it already has a dtype, which also applies to the missing values and we should respect, so any kind of inference is done on the full array; in practice, because I set it to ``np.array(False)`` precisely to avoid the cost of looking for missing values.

I can:

Maybe I could change the name of the variable: ``null_mask`` is not "the mask of null values" but it is "the mask of values we want to leave out for the inference steps", and the two are the same only when ``values`` is a list.

Reply:
That is: unless you have in mind future refactorings for which the location of missing values in an array matters.

Reply:
can you make the change as discussed

Reply:
IOW compute null_mask at the top

Reply:
> can you make the change as discussed

As I explained [above](https://github.com/pandas-dev/pandas/pull/17934#discussion_r146851154), computing ``null_mask`` at the top is a waste, we don't need it for arrays/ndframes/indexes, as ``factorize`` looks for missing values anyway.